source/extensions/filters/http/composite/action.cc

```diff
@@ -66,6 +131,31 @@ Matcher::ActionFactoryCb ExecuteFilterActionFactory::createDynamicActionFactoryC
       provider_manager);
 }
 
+Matcher::ActionFactoryCb ExecuteFilterActionFactory::createAtionFactoryCbCommon(
+    const envoy::extensions::filters::http::composite::v3::ExecuteFilterAction& composite_action,
+    Http::Matching::HttpFilterActionContext& context, Envoy::Http::FilterFactoryCb& callback,
+    bool is_downstream) {
+  const std::string stream_str = is_downstream ? "downstream" : "upstream";
+
+  if (callback == nullptr) {
+    throw EnvoyException(
+        fmt::format("Failed to get {} filter factory creation function", stream_str));
+  }
+  std::string name = composite_action.typed_config().name();
+  ASSERT(context.server_factory_context_ != absl::nullopt);
+  Random::RandomGenerator& random = context.server_factory_context_->api().randomGenerator();
+  Envoy::Runtime::Loader& runtime = context.server_factory_context_->runtime();
+
+  return [cb = std::move(callback), n = std::move(name),
+          composite_action = std::move(composite_action), &random, &runtime,
+          this]() -> Matcher::ActionPtr {
+    if (isSampled(composite_action, random, runtime)) {
```

Comment:
The `isSampled` function is called on the dataplane threads. As such it can not throw C++ exceptions. But the getSampleRatio that is called by isSampled throws exceptions.

Also note that `getSampleRatio` parses the configuration on every sample check. Config parsing and error checking should be done once and only runtime value is queried on every sample.

Reply:
Good catch.  Separated the configuration parsing (getSampleRatio()), and the runtime key retrieving code(getSampleRatioRuntime()). getSampleRatio() is now called in the control plane configuration parsing phase. And isSampled->getSampleRatioRuntime() is called in data plane.